### ISSUE TITLE
feat request: cross-module property accessibility

### ISSUE THREAD

**Issue by trusktr** [microsoft/TypeScript#35554]
## Suggestion

A way to allow certain other modules to access properties or methods of a class.

## Use Cases

Sometimes, we want other modules (other code) to have access to certain properties or methods, but otherwise the end user not to have access to the properties/methods. This is labeled as "package protected" or similar terms in other languages.

## Examples

This is how it could work, with some sort of new non-runtime syntax:

```js
import SomeClass from './SomeClass'

export class Foo {
  visible in SomeClass
  private foo: number = 123
}
```

or

```js
import SomeClass from './SomeClass'

export class Foo {
  private foo visible in SomeClass: number = 123
}
```

and `SomeClass.ts` might have something like:

```js
import {Foo} from './Foo'

export class SomeClass {
  doSomethingWithFoo(o: Foo) {
    console.log(o.foo) // OK, no type error.
  }
}
```

The code inside of `SomeClass` could be allowed to access the private properties. This is great at design time when the author controls the source for both sides, but still wants to hide private parts from the end user on the outside of the APIs.

This allows for patterns like object managers being able to control certain aspects of the objects they manage (for example a renderer in a game engine that manages how objects in a scene render, and the renderable properties that the engine manager accesses, which are private to the end user) are composed from reactions to changes in the objects' public APIs by end users).

At the moment, the only way to allow objects to access properties of other objects is to make them public, but this does not allow for us to hide (at least in the type system) certain implementation details of a cross-class or cross-module whole system from the end user.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [ ] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).
  - I'm not sure about this. This feature wouldn't change semantics of JavaScript, or add any JavaScript features, but would only add additional type capabilities to limit use of APIs in certain ways. In plain JS, before I ported a JavaScript project to TypeScript, I had an object manager that accessed properties prefixed with `__` to denote that the properties were not to be used by the end user. The engine accessed the properties for internal implementation of the system.

**Comment by fatcerberus** [comment]
Seems like a duplicate of https://github.com/microsoft/TypeScript/issues/5228 (C#-style `internal`), although your actual proposal looks semantically closer to C++‘s `friend`.

**Comment by trusktr** [comment]
Ah, thanks for linking me to that. Yeah, this is totally closer to C++ `friend` than C# `internal` or Java "package protected".

I think this is more of an alternative than a duplicate, because `internal` (like C#) is not so useful in a language without any concept of a unit of compilation built from more than one file. Basically what I described at https://github.com/microsoft/TypeScript/issues/5228#issuecomment-562822307.

**Comment by lorenzodallavecchia** [comment]
Since we are listing similar features in other languages, there are also these two:

* [InternalsVisibleTo](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute?redirectedfrom=MSDN&view=netframework-4.8) for sharing internal visibility between assemblies in .NET.

* [@package](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#package) annotation in Closure Compiler JSDoc, which retricts visibility to a directory. One of the few things I miss from the pre-TypeScript days.

**Comment by fatcerberus** [comment]
One problem with a declaration like `@visibleIn(SomeClass)` - what is `SomeClass`?  This question is 100% clear in a language like C++ where all types are nominal, but TypeScript is structurally typed.  So are we saying that all types *shaped like* `SomeClass` are friends, or all types *named* `SomeClass` are?  Neither scenario is ideal IMO, and `SomeClass` only becomes a truly nominal type (the only ideal scenario) if it has `private` members.

**Comment by trusktr** [comment]
> One problem with a declaration like @visibleIn(SomeClass) - what is SomeClass?

In this case, SomeClass is in particular the exact class exported from the module where it is imported from. It isn't regarded as any sort of structural type, but merely tells the compiler that any code in SomeClass can access private members of instances of the current class. I guess it isn't about structural typing at all in this sense.

In some sense, it refers to SomeClass nominally, but SomeClass isn't being type checked here.

**Comment by lorenzodallavecchia** [comment]
Maybe it could be useful to think of friend visibility at _module_ level instead of classes. Modules are a native encapsulation concept in JavaScript, so it would probably feel more natural to reason on them, especially for the many folks that avoid classes in their programming style.

**Comment by trusktr** [comment]
Yeah, totally. Maybe we can have both? Because what if I have more than one class in a module, but I want the member visible only in a specific class?

What about similar to above, but referring to a module:

```ts
export class Foo {
  visible in import('./SomeClass')
  private foo: number = 123
}
```

or

```ts
type Friend = typeof import('./SomeClass')

export class Foo {
  visible in Friend
  private foo: number = 123
}
```

or something?

Any other syntax ideas?

**Comment by EKashpersky** [comment]
IMO it makes sense to have `visible` keyword as a replacement to `export` (in some cases of course) like so:

**Foo.ts**

```typescript
namespace NS {
  visible class Foo {}
}
```

**Bar.ts**

```typescript
namespace NS {
  export class PublicAPI {
    public constructor () {
      const tmp = new NS.Foo();
    }
  }
}
```

**index.ts**

```typescript
const API = new NS.PublicAPI();

new NS.Foo(); // Cannot find class Foo within NS namespace
```

**Reference tags omitted intentionally.**

So if something is `visible` then we can access it within same namespace(for ex.), if `export` then we can access it outside.

Of course, it's a bit off the topic, I understand.
Still I think it'd allow to have same kind of functionality and having relatively small changes to the syntax.

**Comment by captainrdubb** [comment]
Please, in the name of the Common-Closure Principle, add this feature. Doesn't this align with the spirit of Typescript?

**Comment by branko-d** [comment]
Folder or file-level visibility or even a C++ style `friend` mechanism would be great for some types of React components.

In my case, I have a MobX-observable _model_ class which drives the rendering of a purely-functional React _component_. Some of the observables in the model are an "implementation detail" necessary for the functioning of the component, while others represent its "public interface". It would be great if I could make the former visible to the component only, and latter visible to everyone.

A typical example would be a picker: it can render an entire list of pickable items ("implementation detail"), and the consumer reads the currently picked item ("public interface"). The model needs both to properly "drive" the component, but the outside world will only ever be interested in the latter.

**Comment by hinell** [comment]
No need for this madness.

**Comment by TurboEncabulator9000** [comment]
It would be nice to have this available for either module- or class-style development.

Maybe someone who never wrote a line of code until one day they went to a boot camp, and came out the other end a JS developer, and never wrote a single line in another language, thinks classes are dumb. However, a lot of us come from developing in various different OOP languages, and don't want to write blobs of functions with global variables floating around. We already have years/decades of experience before we start messing with JS/TS. We don't want to throw that away and lose all of the value it brings, and which TS _already_ works quite hard to give us. (Really, what experienced OOP developer would want to write anything serious in plain JS if they didn't have to, and if TS wasn't around?)

If JS is going to be the "eight ways of doing everything" language, and if those who maintain it don't care to constrain it from that, and now there are essentially two distinct development methodologies, then why say only one should get this feature?

The more preeminent proposal is the `internal` keyword, but this doesn't prevent someone working _within_ a repo who doesn't understand the intended architecture from misusing a public API.

For example, suppose I have these layers:
* Controller
* Service
* Adapter (factoried by the Service)

Nice and clean, easy to test, SOLID-compliant.

Service supplies some methods that would be useful by the Adapter, but it would be totally wrong for the Controller to call those methods. Therefore, only code "below" the Service (that is, in the Adapters) should call it.

However, suppose you have some developers who can't understand that. They see a new use case, and want to tack on some conditionals in the Controller that reach into the Service and call its public methods, usurping responsibility from the Adapters. You can't control what they do because they don't all send their PRs to you, and those who will review those PRs don't get the layered design either.

You _want_ to express this intended design using a "friend" mechanism, but you can't. Next option is `internal`, which is irrelevant to the problem at hand.

What are your alternatives? Either make the Service an abstract class which would be extended by an Adapter (in which case it becomes quite sprawling, and violates the Single Responsibility Principle), or you make several of its methods public and tolerate the risk that someone is going to do something goofy with them from a Controller.

**Comment by azerum** [comment]
@TurboEncabulator9000 In your example, is this possible to define two interfaces that Service implements - one for methods exposed to Controller, and other for methods exposed to Adapters, and then, instead of making Controller and Adapters interact with the service directly, make them depend on interfaces?
